File: xmake/scripts/winmd/appsdk.py
#!/usr/bin/env python3
"""Windows App SDK WinMD collection module.

从 Windows App SDK 2.0.1+ 的三个子包中自动发现并收集 WinMD 元数据文件：
Microsoft.WindowsAppSDK.Foundation、Microsoft.WindowsAppSDK.WinUI、
Microsoft.WindowsAppSDK.InteractiveExperiences。

采用 glob 方式自动发现 metadata/ 目录下的所有 *.winmd 文件，
无需维护硬编码文件名清单，兼容未来 WinMD 增删。
"""

from __future__ import annotations

from pathlib import Path

from plat_info import BuildError, parse_version, require_dir
from nuget_config import NuGetConfig


# ── 统一收集接口 ──────────────────────────────────────────────


def collect(config: NuGetConfig) -> list[Path]:
    """从 WinAppSDK 子包元数据目录中自动收集 WinMD 文件。

    Args:
        config: 已解析的 NuGet 包配置实例。

    Returns:
        所有 WinAppSDK WinMD 文件的 Path 列表（按文件名排序）。
    """
    winmds: list[Path] = []

    # Foundation 子包 — 根级别 metadata/*.winmd
    foundation_pkg = config.package_path("Microsoft.WindowsAppSDK.Foundation")
    foundation_meta = require_dir(
        foundation_pkg / "metadata", "Foundation metadata directory"
    )
    winmds.extend(sorted(foundation_meta.glob("*.winmd")))

    # WinUI 子包 — 根级别 metadata/*.winmd
    winui_pkg = config.package_path("Microsoft.WindowsAppSDK.WinUI")
    winui_meta = require_dir(winui_pkg / "metadata", "WinUI metadata directory")
    winmds.extend(sorted(winui_meta.glob("*.winmd")))

    # InteractiveExperiences 子包 — 版本化 metadata/<max_version>/*.winmd
    ixp_pkg = config.package_path("Microsoft.WindowsAppSDK.InteractiveExperiences")
    ixp_meta_root = require_dir(
        ixp_pkg / "metadata", "InteractiveExperiences metadata directory"
    )
    ixp_versions = sorted(
        (d.name for d in ixp_meta_root.iterdir() if d.is_dir()),
        key=parse_version,
    )
    if not ixp_versions:
        raise BuildError(f"no version sub-directories in: {str(ixp_meta_root)}")
    ixp_meta = require_dir(
        ixp_meta_root / ixp_versions[-1], "InteractiveExperiences version metadata"
    )
    winmds.extend(sorted(ixp_meta.glob("*.winmd")))

    return winmds

File: xmake/modules/winui3/winmd/appsdk.lua
-- 从 Windows App SDK 2.0.1+ 的三个子包中收集 WinMD 元数据文件：
--   Microsoft.WindowsAppSDK.Foundation          — metadata/*.winmd
--   Microsoft.WindowsAppSDK.WinUI               — metadata/*.winmd
--   Microsoft.WindowsAppSDK.InteractiveExperiences — metadata/<version>/*.winmd

local packages = import("winui3.packages")
local utils    = import("utils")

-- InteractiveExperiences 使用版本化 metadata 子目录，需选择最高版本目录。
local function _find_max_version_subdir(dir_path)
    if not os.isdir(dir_path) then
        return nil
    end

    local names   = {}
    local entries = {}
    for _, entry in ipairs(os.dirs(path.join(dir_path, "*"))) do
        local name = path.filename(entry)
        names[#names + 1] = name
        entries[name] = entry
    end
    local best_name = utils.max_version(names)
    return best_name and entries[best_name] or nil
end

function collect()
    local foundation_root = packages.package_root("Microsoft.WindowsAppSDK.Foundation")
    local foundation_meta = path.join(foundation_root, "metadata")
    if not os.isdir(foundation_meta) then
        raise(string.format("Foundation metadata 目录不存在：%s", foundation_meta))
    end
    local fnd_winmds = os.files(path.join(foundation_meta, "*.winmd"))
    if #fnd_winmds == 0 then
        raise(string.format("Foundation metadata 目录中无 .winmd 文件：%s", foundation_meta))
    end
    local winui_root = packages.package_root("Microsoft.WindowsAppSDK.WinUI")
    local winui_meta = path.join(winui_root, "metadata")
    if not os.isdir(winui_meta) then
        raise(string.format("WinUI metadata 目录不存在：%s", winui_meta))
    end
    local wui_winmds = os.files(path.join(winui_meta, "*.winmd"))
    if #wui_winmds == 0 then
        raise(string.format("WinUI metadata 目录中无 .winmd 文件：%s", winui_meta))
    end

    local ixp_root = packages.package_root("Microsoft.WindowsAppSDK.InteractiveExperiences")
    local ixp_meta_root = path.join(ixp_root, "metadata")
    if not os.isdir(ixp_meta_root) then
        raise(string.format("InteractiveExperiences metadata 目录不存在：%s", ixp_meta_root))
    end
    local ixp_version_dir = _find_max_version_subdir(ixp_meta_root)
    if not ixp_version_dir then
        raise(string.format("InteractiveExperiences metadata 下无版本子目录：%s", ixp_meta_root))
    end
    local ixp_winmds = os.files(path.join(ixp_version_dir, "*.winmd"))
    if #ixp_winmds == 0 then
        raise(string.format(
            "InteractiveExperiences metadata 版本目录中无 .winmd 文件：%s", ixp_version_dir))
    end

    local winmds = table.join(fnd_winmds, wui_winmds, ixp_winmds)
    return winmds
end

